File: scripts/backend-python/webhook_handler.py
"""Paytm S2S webhook receiver.

See backend-node/webhookHandler.js for the same contract notes — production
expectations are identical (Redis dedup, IP whitelist, 5xx on processing
errors, fast 200 with heavy work queued).
"""
import re
import threading
from collections import deque
from datetime import datetime, timezone
from typing import Any, Optional, Tuple

from paytmchecksum import PaytmChecksum

from paytm_config import get_paytm_config

_seen: set = set()                       # "{orderId}|{status}" — at-least-once dedup
_SEEN_MAX = 50_000
_event_log = deque(maxlen=200)
_lock = threading.Lock()


def recent_events():
    with _lock:
        return list(_event_log)[::-1]


def _extract_body_bytes(raw_body: str) -> Optional[str]:
    """Substring of the raw body that corresponds to "body": {...}.

    Paytm signs those bytes verbatim — re-serializing here would change key
    order / whitespace and break the signature.
    """
    if not raw_body:
        return None
    m = re.search(r'"body"\s*:\s*\{', raw_body)
    if not m:
        return None
    start = m.end() - 1                 # position of the opening '{'
    depth = 0
    in_string = False
    escape = False
    for i in range(start, len(raw_body)):
        c = raw_body[i]
        if escape:
            escape = False
            continue
        if c == "\\":
            escape = True
            continue
        if c == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if c == "{":
            depth += 1
        elif c == "}":
            depth -= 1
            if depth == 0:
                return raw_body[start:i + 1]
    return None


def handle_webhook(raw_body: str, parsed: Any) -> Tuple[int, dict]:
    cfg = get_paytm_config()
    if not cfg["merchant_key"]:
        return 500, {"ok": False, "error": "merchant key not configured"}

    signature = ((parsed or {}).get("head") or {}).get("signature")
    if not signature:
        return 401, {"ok": False, "error": "missing head.signature"}

    body_bytes = _extract_body_bytes(raw_body)
    if body_bytes is None:
        # Fallback — but this is best-effort and may fail signature check.
        import json
        body_bytes = json.dumps((parsed or {}).get("body") or {})

    try:
        signature_ok = PaytmChecksum.verifySignature(body_bytes, cfg["merchant_key"], signature)
    except Exception:
        signature_ok = False
    if not signature_ok:
        return 401, {"ok": False, "error": "invalid signature"}

    # Persist for audit BEFORE dedup so duplicates are visible.
    with _lock:
        _event_log.append({
            "at": datetime.now(timezone.utc).isoformat(),
            "payload": parsed,
        })

    body_obj = (parsed or {}).get("body") or {}
    order_id = body_obj.get("orderId")
    status = body_obj.get("status") or body_obj.get("STATUS")
    dedup_key = f"{order_id or 'unknown'}|{status or 'unknown'}"

    with _lock:
        if dedup_key in _seen:
            return 200, {"ok": True, "dedup": True, "orderId": order_id, "status": status}
        if len(_seen) >= _SEEN_MAX:
            _seen.clear()
        _seen.add(dedup_key)

    _fulfill_order(order_id, status, parsed)
    return 200, {"ok": True, "orderId": order_id, "status": status}


def _fulfill_order(order_id, status, parsed):
    """Replace with your real DB write / queue push. Keep it FAST — webhook
    timeout is 10s; queue heavy work (emails, accounting) if needed."""
    print("[paytm webhook] fulfill stub", {
        "orderId": order_id,
        "status": status,
        "mid": ((parsed or {}).get("body") or {}).get("mid"),
    })

File: scripts/backend-node/webhookHandler.js
// Paytm S2S webhook receiver.
//
// Contract:
//   1. Read raw body bytes (we capture them in server.js via express.json verify).
//   2. Extract head.signature; verify against the body bytes Paytm signed.
//      Re-serializing here would change key order / whitespace and break the
//      signature — use the bytes Paytm sent.
//   3. Idempotency check: (orderId, status) — Paytm retries at-least-once.
//   4. Persist event for audit (here, just an in-memory log).
//   5. Apply state transition (here, a stub `fulfillOrder` hook — replace with
//      your real DB write).
//   6. Return 200 fast. Heavy lifting (emails, accounting) goes to a queue.
//
// PRODUCTION NOTES:
//   - Replace the in-memory dedup set with Redis / a DB unique-index on
//     (orderId, status).
//   - Whitelist Paytm's egress IPs at your firewall (request the current list
//     from Paytm support).
//   - Return 5xx on processing errors so Paytm retries; do NOT return 200 on
//     handler crashes.
import PaytmChecksum from "paytmchecksum";
import { getPaytmConfig } from "./paytmConfig.js";

const seen = new Set();                       // `${orderId}|${status}` — at-least-once dedup
const SEEN_MAX = 50_000;
const eventLog = [];                          // ring buffer for /paytm/webhook/events
const EVENT_LOG_MAX = 200;

function rememberEvent(parsed) {
  eventLog.push({ at: new Date().toISOString(), payload: parsed });
  if (eventLog.length > EVENT_LOG_MAX) eventLog.shift();
}

export function recentEvents() {
  return eventLog.slice().reverse();
}

/**
 * Substring of the raw body that corresponds to "body": {...}.
 * Paytm signs those bytes verbatim — re-serializing the parsed object would
 * change key order / whitespace and break the signature.
 */
function extractBodyBytes(rawBody) {
  if (!rawBody) return null;
  // Find the literal substring `"body":` then walk braces.
  const m = rawBody.match(/"body"\s*:\s*\{/);
  if (!m) return null;
  const start = m.index + m[0].length - 1;     // position of the opening '{'
  let depth = 0;
  let inString = false;
  let escape = false;
  for (let i = start; i < rawBody.length; i++) {
    const c = rawBody[i];
    if (escape) { escape = false; continue; }
    if (c === "\\") { escape = true; continue; }
    if (c === '"') { inString = !inString; continue; }
    if (inString) continue;
    if (c === "{") depth++;
    else if (c === "}") {
      depth--;
      if (depth === 0) return rawBody.slice(start, i + 1);
    }
  }
  return null;
}

export async function handleWebhook({ rawBody, parsed }) {
  const cfg = getPaytmConfig();
  if (!cfg.merchantKey) {
    return { ok: false, httpStatus: 500, detail: { error: "merchant key not configured" } };
  }

  const signature = parsed?.head?.signature;
  if (!signature) {
    return { ok: false, httpStatus: 401, detail: { error: "missing head.signature" } };
  }

  // Verify the signed body bytes (NOT a re-serialization).
  const bodyBytes = extractBodyBytes(rawBody) || JSON.stringify(parsed?.body || {});
  let signatureOk = false;
  try {
    signatureOk = PaytmChecksum.verifySignature(bodyBytes, cfg.merchantKey, signature);
  } catch {
    signatureOk = false;
  }
  if (!signatureOk) {
    return { ok: false, httpStatus: 401, detail: { error: "invalid signature" } };
  }

  // Persist for audit BEFORE dedup so we can see duplicates if they arrive.
  rememberEvent(parsed);

  // Idempotency: at-least-once delivery means duplicates are normal.
  // Key on (orderId, status) — Paytm always sets both for terminal events.
  const orderId = parsed?.body?.orderId;
  const status = parsed?.body?.status || parsed?.body?.STATUS;
  const dedupKey = `${orderId || "unknown"}|${status || "unknown"}`;
  if (seen.has(dedupKey)) {
    return { ok: true, httpStatus: 200, detail: { dedup: true, orderId, status } };
  }
  if (seen.size >= SEEN_MAX) seen.clear();   // crude wrap; Redis has TTLs in real life
  seen.add(dedupKey);

  // Apply state transition. Stub here — replace with your real fulfillment.
  await fulfillOrder({ orderId, status, parsed });

  return { ok: true, httpStatus: 200, detail: { orderId, status } };
}

/**
 * Replace this with your real DB write / queue push. Keep it FAST — webhook
 * timeout is 10s; queue heavy work (emails, accounting) if needed.
 */
async function fulfillOrder({ orderId, status, parsed }) {
  // eslint-disable-next-line no-console
  console.log("[paytm webhook] fulfill stub", { orderId, status, mid: parsed?.body?.mid });
}
